rdata.c

```diff
@@ -845,515 +810,3566 @@ rdata_svcparam_alpn_to_string(buffer_type *output, uint16_t val_len,
+		       (unsigned long) domain->number,
+		       query_get_dname_offset(q, domain)));
+		buffer_write(q->packet, dname_name(domain_dname(domain)),
+			     label_length(dname_name(domain_dname(domain))) + 1U);
+		domain = domain->parent;
+	}
+	if (domain->parent) {
+		DEBUG(DEBUG_NAME_COMPRESSION, 2,
+		      (LOG_INFO, "dname: %s, number: %lu, pointer: %u\n",
+		       domain_to_string(domain),
+		       (unsigned long) domain->number,
+		       query_get_dname_offset(q, domain)));
+		assert(query_get_dname_offset(q, domain) <= MAX_COMPRESSION_OFFSET);
+		buffer_write_u16(q->packet,
+				 0xc000 | query_get_dname_offset(q, domain));
+	} else {
+		buffer_write_u8(q->packet, 0);
+	}
+}
+
+void
+write_compressed_name_rdata(struct query *query, const struct rr *rr)
+{
+	struct domain *domain;
+	assert(rr->rdlength == sizeof(void*));
+	memcpy(&domain, rr->rdata, sizeof(void*));
+	encode_dname(query, domain);
+}
+
+void
+write_uncompressed_name_rdata(struct query *query, const struct rr *rr)
+{
+	const struct dname *dname;
+	struct domain *domain;
+	assert(rr->rdlength >= sizeof(void*));
+	memcpy(&domain, rr->rdata, sizeof(void*));
+	dname = domain_dname(domain);
+	buffer_write(query->packet, dname_name(dname), dname->name_size);
+}
+
+int
+print_name_rdata(struct buffer *output, const struct rr *rr)
+{
+	uint16_t length = 0;
+	assert(rr->rdlength == sizeof(void*));
+	/* This prints a reference to a name stored as a pointer. */
+	return print_domain(output, rr->rdlength, rr->rdata, &length);
+}
+
+int32_t
+read_a_rdata(struct domain_table *domains, uint16_t rdlength,
+	struct buffer *packet, struct rr **rr)
+{
+	if (rdlength != 4)
+		return MALFORMED;
+	return read_rdata(domains, rdlength, packet, rr);
+}
+
+int
+print_a_rdata(struct buffer *output, const struct rr *rr)
+{
+	uint16_t length = 0;
+	assert(rr->rdlength == 4);
+	return print_ip4(output, rr->rdlength, rr->rdata, &length);
+}
+
+int32_t
+read_soa_rdata(struct domain_table *domains, uint16_t rdlength,
+	struct buffer *packet, struct rr **rr)
+{
+	struct domain *primary_domain, *mailbox_domain;
+	struct dname_buffer primary, mailbox;
+	size_t size;
+
+	/* domain + domain + uint32 + uint32 + uint32 + uint32 + uint32 */
+	const size_t mark = buffer_position(packet);
+	if (!dname_make_from_packet_buffered(&primary, packet, 1, 1) ||
+	    !dname_make_from_packet_buffered(&mailbox, packet, 1, 1) ||
+	    rdlength != (buffer_position(packet) - mark) + 20)
+		return MALFORMED;
+
+	size = sizeof(**rr) + 2 * sizeof(struct domain *) + 20;
+	if (!(*rr = region_alloc(domains->region, size)))
+		return TRUNCATED;
+	primary_domain = domain_table_insert(domains, (void*)&primary);
+	primary_domain->usage++;
+	mailbox_domain = domain_table_insert(domains, (void*)&mailbox);
+	mailbox_domain->usage++;
+
+	memcpy((*rr)->rdata, &primary_domain, sizeof(void*));
+	memcpy((*rr)->rdata + sizeof(void*), &mailbox_domain, sizeof(void*));
+	buffer_read(packet, (*rr)->rdata + 2 * sizeof(void*), 20);
```

Comment:
While that is unlikely to ever be an issue, do you want to keep it consistent with the other sizeof pointers calculations?

```suggestion
	size = sizeof(**rr) + 2 * sizeof(void*) + 20;
	if (!(*rr = region_alloc(domains->region, size)))
		return TRUNCATED;
	primary_domain = domain_table_insert(domains, (void*)&primary);
	primary_domain->usage++;
	mailbox_domain = domain_table_insert(domains, (void*)&mailbox);
	mailbox_domain->usage++;

	memcpy((*rr)->rdata, &primary_domain, sizeof(void*));
	memcpy((*rr)->rdata + sizeof(void*), &mailbox_domain, sizeof(void*));
	buffer_read(packet, (*rr)->rdata + 2 * sizeof(void*), 20);
```

Reply:
It is fine to have the change. The consistency is nice to have.